third_party/amd/lib/TritonAMDGPUTransforms/WarpPipeliner.cpp

```diff
@@ -0,0 +1,181 @@
+        return !opsLookup.count(u) && u->getParentOp() != nullptr;
+      });
+      if (hasExternalUse) {
+        yieldedTypes.push_back(r.getType());
+        resultToYieldIdx.emplace_back(
+            r, static_cast<unsigned>(yieldedTypes.size() - 1));
+      }
+    }
+  }
+
+  // Create the execute_region with the final result types.
+  auto exec = b.create<scf::ExecuteRegionOp>(loc, yieldedTypes);
+  Block *body = &exec.getRegion().emplaceBlock();
+  b.setInsertionPointToStart(body);
+
+  // Clone ops in order, remapping intra-cluster defs to their clones.
+  IRMapping mapping;
+  for (Operation *op : ops) {
+    Operation *clone = b.clone(*op, mapping);
+    // Map each result so subsequent clones use the cloned defs.
+    for (auto [origRes, clonedRes] :
+         llvm::zip(op->getResults(), clone->getResults()))
+      mapping.map(origRes, clonedRes);
+  }
+
+  // Build the yield values.
+  SmallVector<Value> yieldVals(yieldedTypes.size());
+  for (auto [origRes, yieldIdx] : resultToYieldIdx) {
+    Value mapped = mapping.lookupOrNull(origRes);
+    assert(mapped && "mapped result missing");
+    yieldVals[yieldIdx] = mapped;
+  }
+  b.create<scf::YieldOp>(loc, yieldVals);
+
+  // Replace external uses of original results with exec results.
+  // Internal uses were already remapped when cloning.
+  for (auto [origRes, yieldIdx] : resultToYieldIdx) {
+    Value repl = exec.getResult(yieldIdx);
+    origRes.replaceUsesWithIf(repl, [&](OpOperand &use) {
+      Operation *owner = use.getOwner();
+      return owner && owner->getParentOp() != exec;
+    });
+  }
+
+  // Erase original ops now that their external uses are redirected.
+  std::reverse(ops.begin(), ops.end());
+  for (Operation *op : ops)
+    op->erase();
+
+  // Keep the region structured for later conversion.
+  exec.setNoInline(true);
+  exec->setAttr("triton.warp_pipeline.stage", b.getUnitAttr());
+
+  LLVM_DEBUG(llvm::dbgs() << "[warp-pipeline] created stage with " << ops.size()
+                          << " ops and " << yieldedTypes.size() << " yields\n");
+  return;
+}
+
+// Turns a partitioned region into the warp-pipelined clusters
+static LogicalResult createPipeline(OpBuilder &b, Location loc,
+                                    scf::ForOp forOp) {
+  // Collect ops in the loop body
+  Block &blk = *forOp.getBody();
+  SmallVector<Operation *> cluster;
+  SmallVector<SmallVector<Operation *>> clusters;
+
+  // ops cannot be located within a cluster
+  // barrier/wait still require border op
+  auto isIgnorable = [](Operation *op) {
+    return isa<scf::ForOp, scf::YieldOp, ttg::AsyncWaitOp, gpu::BarrierOp,
+               tt::amdgpu::AsyncTDMWait>(op);
+  };
+
+  auto isBorder = [](Operation *op) {
+    return op->hasAttr("triton.warp_pipeline.border");
+  };
+
+  // One pass over the body; collect clusters split by explicit borders.
+  for (Operation &opRef : llvm::make_early_inc_range(blk)) {
+    Operation *op = &opRef;
+    if (isIgnorable(op))
```

Comment:
I'm not sure I fully follow the logic here--it ignores these ops at the beginning before any cluster but later will be put into the cluster? Can you probably add some comments in the code to explain? We don't have lit tests covering this part?